File: ios/MinesweeperApp/MinesweeperApp.swift
import SwiftUI
import MinesweeperCore

@main
struct MinesweeperApp: App {
    var body: some Scene {
        WindowGroup {
            ContentView()
                .preferredColorScheme(.dark)
        }
    }
}

struct Theme {
    static let bg = Color(red: 12/255, green: 16/255, blue: 36/255)
    static let panel = Color(red: 22/255, green: 26/255, blue: 58/255)
    static let panel2 = Color(red: 31/255, green: 36/255, blue: 82/255)
    static let ink = Color(red: 233/255, green: 236/255, blue: 255/255)
    static let muted = Color(red: 141/255, green: 146/255, blue: 199/255)
    static let accent = Color(red: 76/255, green: 201/255, blue: 240/255)
    static let accent2 = Color(red: 179/255, green: 136/255, blue: 255/255)
    static let danger = Color(red: 255/255, green: 92/255, blue: 141/255)
    static let good = Color(red: 110/255, green: 240/255, blue: 163/255)
    static let hidden = Color(red: 53/255, green: 58/255, blue: 120/255)
    static let revealed = Color(red: 15/255, green: 19/255, blue: 48/255)

    static func numberColor(_ n: Int) -> Color {
        switch n {
        case 1: return accent
        case 2: return good
        case 3: return danger
        case 4: return accent2
        case 5: return Color(red: 1, green: 180/255, blue: 84/255)
        case 6: return Color(red: 74/255, green: 215/255, blue: 209/255)
        case 7: return ink
        default: return muted
        }
    }
}

@MainActor
final class GameModel: ObservableObject {
    @Published private(set) var game: Game
    @Published var preset: Preset = .beginner
    @Published var flagMode = false
    @Published var tick = 0

    private var timer: Timer?

    init() {
        game = Game(.beginner)
        timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.tick &+= 1 }
        }
    }

    func newGame(_ p: Preset) {
        preset = p
        game = Game(p)
    }

    func tap(_ r: Int, _ c: Int) {
        let cell = game.cells[r][c]
        if flagMode {
            game.toggleFlag(r, c)
        } else if cell.state == .revealed && cell.adjacent > 0 {
            game.chord(r, c)
        } else {
            game.reveal(r, c)
        }
        objectWillChange.send()
    }

    func longPress(_ r: Int, _ c: Int) {
        game.toggleFlag(r, c)
        objectWillChange.send()
    }
}

struct ContentView: View {
    @StateObject private var model = GameModel()

    var body: some View {
        ZStack {
            LinearGradient(colors: [Theme.bg, Theme.panel], startPoint: .top, endPoint: .bottom)
                .ignoresSafeArea()

            VStack(spacing: 14) {
                Text("MINE·SWEEPER · iOS")
                    .font(.caption.bold())
                    .foregroundColor(Theme.accent)
                    .padding(.top, 4)

                hud

                presets

                modeToggle

                ScrollView([.horizontal, .vertical]) {
                    board
                        .padding(8)
                        .background(Color(red: 27/255, green: 32/255, blue: 80/255))
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                }

                Text("Tap reveal · Long-press flag · Tap a number to chord")
                    .font(.caption2).foregroundColor(Theme.muted)
            }
            .padding(.horizontal, 16)

            if model.game.status == .won || model.game.status == .lost {
                banner
            }
        }
    }

    private var hud: some View {
        HStack {
            Text(String(format: "%03d", model.game.minesRemaining))
                .font(.system(.title, design: .monospaced).bold())
                .foregroundColor(Theme.danger)
                .padding(.horizontal, 12).padding(.vertical, 4)
                .background(Color(red: 16/255, green: 0, blue: 32/255))
                .clipShape(RoundedRectangle(cornerRadius: 8))
            Spacer()
            Button(action: { model.newGame(model.preset) }) {
                Text(face).font(.title)
            }
            .frame(width: 52, height: 52)
            .background(Color(red: 245/255, green: 197/255, blue: 24/255))
            .clipShape(Circle())
            Spacer()
            Text(String(format: "%03d", model.game.elapsedSeconds))
                .font(.system(.title, design: .monospaced).bold())
                .foregroundColor(Theme.accent)
                .padding(.horizontal, 12).padding(.vertical, 4)
                .background(Color(red: 16/255, green: 0, blue: 32/255))
                .clipShape(RoundedRectangle(cornerRadius: 8))
        }
        .padding(12)
        .background(Theme.panel)
        .clipShape(RoundedRectangle(cornerRadius: 14))
    }

    private var face: String {
        switch model.game.status {
        case .lost: return "😵"
        case .won: return "😎"
        default: return "🙂"
        }
    }

    private var presets: some View {
        HStack(spacing: 8) {
            ForEach(Preset.all, id: \.name) { p in
                Button(p.name) { model.newGame(p) }
                    .font(.subheadline.bold())
                    .padding(.horizontal, 16).padding(.vertical, 8)
                    .background(p == model.preset ? Theme.accent : Theme.panel2)
                    .foregroundColor(p == model.preset ? .black : Theme.ink)
                    .clipShape(Capsule())
            }
        }
    }

    private var modeToggle: some View {
        Button(action: { model.flagMode.toggle() }) {
            Label(model.flagMode ? "Flag mode (on)" : "Flag mode",
                  systemImage: "flag.fill")
                .font(.footnote.bold())
                .padding(.horizontal, 14).padding(.vertical, 6)
                .background(model.flagMode ? Theme.accent2 : Theme.panel2)
                .foregroundColor(model.flagMode ? .black : Theme.ink)
                .clipShape(Capsule())
        }
    }

    private var board: some View {
        VStack(spacing: 2) {
            ForEach(0..<model.game.rows, id: \.self) { r in
                HStack(spacing: 2) {
                    ForEach(0..<model.game.cols, id: \.self) { c in
                        cellView(r: r, c: c)
                    }
                }
            }
        }
    }

    private func cellView(r: Int, c: Int) -> some View {
        let cell = model.game.cells[r][c]
        let size: CGFloat = model.game.cols > 16 ? 24 : 30
        return ZStack {
            RoundedRectangle(cornerRadius: 4)
                .fill(cellBg(cell))
            Text(label(cell))
                .font(.system(size: size * 0.55, weight: .heavy, design: .monospaced))
                .foregroundColor(textColor(cell))
        }
        .frame(width: size, height: size)
        .onTapGesture { model.tap(r, c) }
        .onLongPressGesture(minimumDuration: 0.28) { model.longPress(r, c) }
    }

    private func cellBg(_ cell: Cell) -> Color {
        if cell.state == .revealed {
            if cell.exploded { return Color(red: 107/255, green: 14/255, blue: 42/255) }
            if cell.mine { return Color(red: 42/255, green: 15/255, blue: 28/255) }
            return Theme.revealed
        }
        return Theme.hidden
    }

    private func label(_ cell: Cell) -> String {
        if cell.wrongFlag { return "✕" }
        switch cell.state {
        case .flagged: return "⚑"
        case .questioned: return "?"
        case .revealed:
            if cell.mine { return "✱" }
            return cell.adjacent > 0 ? "\(cell.adjacent)" : ""
        case .hidden: return ""
        }
    }

    private func textColor(_ cell: Cell) -> Color {
        if cell.wrongFlag { return Theme.danger }
        switch cell.state {
        case .flagged: return Theme.accent2
        case .questioned: return Color(red: 1, green: 180/255, blue: 84/255)
        case .revealed:
            if cell.mine { return Theme.danger }
            return Theme.numberColor(cell.adjacent)
        case .hidden: return Theme.ink
        }
    }

    private var banner: some View {
        ZStack {
            Color.black.opacity(0.6).ignoresSafeArea()
            VStack(spacing: 12) {
                Text(model.game.status == .won ? "✨ VICTORY ✨" : "💥 BOOM 💥")
                    .font(.largeTitle.bold())
                    .foregroundColor(model.game.status == .won ? Theme.good : Theme.danger)
                Button("Play again") { model.newGame(model.preset) }
                    .padding(.horizontal, 20).padding(.vertical, 10)
                    .background(Theme.accent)
                    .foregroundColor(.black)
                    .clipShape(Capsule())
            }
        }
    }
}

File: android/app/src/main/java/com/dsalunga/minesweeper/ui/MinesweeperApp.kt
package com.dsalunga.minesweeper.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.dsalunga.minesweeper.Cell
import com.dsalunga.minesweeper.CellState
import com.dsalunga.minesweeper.Game
import com.dsalunga.minesweeper.GameStatus
import com.dsalunga.minesweeper.Preset
import kotlinx.coroutines.delay

private val Bg = Color(0xFF0C1024)
private val Panel = Color(0xFF161A3A)
private val Panel2 = Color(0xFF1F2452)
private val Ink = Color(0xFFE9ECFF)
private val Muted = Color(0xFF8D92C7)
private val Accent = Color(0xFF4CC9F0)
private val Accent2 = Color(0xFFB388FF)
private val Danger = Color(0xFFFF5C8D)
private val Good = Color(0xFF6EF0A3)
private val Hidden = Color(0xFF353A78)
private val Revealed = Color(0xFF0F1330)
private val Mine = Color(0xFF2A0F1C)
private val Exploded = Color(0xFF6B0E2A)

private fun numberColor(n: Int): Color = when (n) {
    1 -> Accent
    2 -> Good
    3 -> Danger
    4 -> Accent2
    5 -> Color(0xFFFFB454)
    6 -> Color(0xFF4AD7D1)
    7 -> Ink
    else -> Muted
}

private val DarkScheme = darkColorScheme(
    primary = Accent, secondary = Accent2, background = Bg, surface = Panel
)

@Composable
fun MinesweeperApp() {
    MaterialTheme(colorScheme = DarkScheme) {
        Surface(color = Bg, modifier = Modifier.fillMaxSize()) {
            GameScreen()
        }
    }
}

@Composable
private fun GameScreen() {
    var preset by remember { mutableStateOf(Preset.Beginner) }
    var game by remember { mutableStateOf(Game(preset)) }
    var version by remember { mutableStateOf(0) }
    var flagMode by remember { mutableStateOf(false) }
    var now by remember { mutableStateOf(System.currentTimeMillis()) }

    LaunchedEffect(Unit) {
        while (true) { now = System.currentTimeMillis(); delay(250) }
    }

    fun bump() { version++ }
    fun reset(p: Preset) { preset = p; game = Game(p); bump() }

    Column(
        Modifier.fillMaxSize().padding(12.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp)
    ) {
        Text("MINE·SWEEPER · ANDROID", color = Accent, fontWeight = FontWeight.Bold, fontSize = 12.sp)

        Hud(game = game, now = now, version = version, onFace = { reset(preset) })

        PresetBar(current = preset, onSelect = { reset(it) })

        FilterChip(
            selected = flagMode,
            onClick = { flagMode = !flagMode },
            label = { Text(if (flagMode) "Flag mode (on)" else "Flag mode") },
            leadingIcon = { Text("⚑") }
        )

        Box(
            Modifier
                .weight(1f)
                .fillMaxWidth()
                .clip(RoundedCornerShape(12.dp))
                .background(Color(0xFF1B2050))
                .padding(8.dp),
            contentAlignment = Alignment.Center
        ) {
            Box(
                Modifier
                    .horizontalScroll(rememberScrollState())
                    .verticalScroll(rememberScrollState())
            ) {
                Board(game = game, version = version, flagMode = flagMode, onChange = { bump() })
            }

            if (game.status == GameStatus.Won || game.status == GameStatus.Lost) {
                Banner(game.status) { reset(preset) }
            }
        }

        Text(
            "Tap reveal · Long-press flag · Tap a number to chord",
            color = Muted, fontSize = 11.sp, textAlign = TextAlign.Center
        )
    }
}

@Composable
private fun Hud(game: Game, now: Long, version: Int, onFace: () -> Unit) {
    @Suppress("UNUSED_EXPRESSION") version  // recompose key
    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(Panel)
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Counter(game.minesRemaining, Danger)
        Spacer(Modifier.weight(1f))
        Box(
            Modifier
                .size(56.dp)
                .clip(CircleShape)
                .background(Color(0xFFF5C518))
                .pointerInput(Unit) {
                    detectTapGestures(onTap = { onFace() })
                },
            contentAlignment = Alignment.Center
        ) {
            Text(
                when (game.status) {
                    GameStatus.Lost -> "X("
                    GameStatus.Won -> "B)"
                    else -> ":)"
                },
                fontWeight = FontWeight.Black, color = Color.Black, fontSize = 20.sp
            )
        }
        Spacer(Modifier.weight(1f))
        Counter(game.elapsedSeconds(now), Accent)
    }
}

@Composable
private fun Counter(value: Int, color: Color) {
    Box(
        Modifier
            .clip(RoundedCornerShape(8.dp))
            .background(Color(0xFF100020))
            .padding(horizontal = 12.dp, vertical = 6.dp)
    ) {
        Text(
            value.toString().padStart(3, '0'),
            color = color,
            fontFamily = FontFamily.Monospace,
            fontWeight = FontWeight.Black,
            fontSize = 22.sp
        )
    }
}

@Composable
private fun PresetBar(current: Preset, onSelect: (Preset) -> Unit) {
    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        Preset.All.forEach { p ->
            val active = p == current
            FilterChip(
                selected = active,
                onClick = { onSelect(p) },
                label = { Text(p.name) },
                colors = FilterChipDefaults.filterChipColors(
                    containerColor = Panel2,
                    selectedContainerColor = Accent,
                    selectedLabelColor = Color.Black
                )
            )
        }
    }
}

@Composable
private fun Board(game: Game, version: Int, flagMode: Boolean, onChange: () -> Unit) {
    @Suppress("UNUSED_EXPRESSION") version
    val size = if (game.cols > 16) 26 else 32
    Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
        for (r in 0 until game.rows) {
            Row(horizontalArrangement = Arrangement.spacedBy(2.dp)) {
                for (c in 0 until game.cols) {
                    CellView(
                        cell = game.cells[r][c], sizeDp = size,
                        onTap = {
                            if (flagMode) game.toggleFlag(r, c)
                            else {
                                val cell = game.cells[r][c]
                                if (cell.state == CellState.Revealed && cell.adjacent > 0) game.chord(r, c)
                                else game.reveal(r, c)
                            }
                            onChange()
                        },
                        onLong = { game.toggleFlag(r, c); onChange() }
                    )
                }
            }
        }
    }
}

@Composable
private fun CellView(cell: Cell, sizeDp: Int, onTap: () -> Unit, onLong: () -> Unit) {
    val bg = when {
        cell.state != CellState.Revealed -> Hidden
        cell.exploded -> Exploded
        cell.mine -> Mine
        else -> Revealed
    }
    val (label, color) = when {
        cell.wrongFlag -> "X" to Danger
        cell.state == CellState.Flagged -> "⚑" to Accent2
        cell.state == CellState.Questioned -> "?" to Color(0xFFFFB454)
        cell.state == CellState.Revealed && cell.mine -> "✱" to Danger
        cell.state == CellState.Revealed && cell.adjacent > 0 -> cell.adjacent.toString() to numberColor(cell.adjacent)
        else -> "" to Ink
    }
    Box(
        Modifier
            .size(sizeDp.dp)
            .clip(RoundedCornerShape(4.dp))
            .background(bg)
            .border(1.dp, Color(0x33000000), RoundedCornerShape(4.dp))
            .pointerInput(cell, label) {
                detectTapGestures(onTap = { onTap() }, onLongPress = { onLong() })
            },
        contentAlignment = Alignment.Center
    ) {
        if (label.isNotEmpty()) {
            Text(
                label, color = color, fontWeight = FontWeight.Black,
                fontFamily = FontFamily.Monospace,
                fontSize = (sizeDp * 0.55).sp
            )
        }
    }
}

@Composable
private fun Banner(status: GameStatus, onAgain: () -> Unit) {
    Box(
        Modifier
            .fillMaxSize()
            .background(Color(0xCC07091F)),
        contentAlignment = Alignment.Center
    ) {
        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Text(
                if (status == GameStatus.Won) "✨ VICTORY ✨" else "💥 BOOM 💥",
                color = if (status == GameStatus.Won) Good else Danger,
                fontWeight = FontWeight.Black, fontSize = 32.sp
            )
            Button(onClick = onAgain, colors = ButtonDefaults.buttonColors(containerColor = Accent, contentColor = Color.Black)) {
                Text("Play again", fontWeight = FontWeight.Bold)
            }
        }
    }
}
